### PR TITLE
refactor: Improve thread safety analysis by propagating some negative capabilities

### PR DESCRIPTION
**Pull request by hebasto**
This PR is a follow-up for bitcoin/bitcoin#22778 and bitcoin/bitcoin#24062, and it seems [required](https://github.com/bitcoin/bitcoin/pull/24931#issuecomment-1132800173) for bitcoin/bitcoin#24931.

See details in the commit messages.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -472,15 +472,16 @@ class PeerManagerImpl final : public PeerManager
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
     void BlockChecked(const CBlock& block, const BlockValidationState& state) override
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
-    void NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock>& pblock) override;
+    void NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock>& pblock) override
+        EXCLUSIVE_LOCKS_REQUIRED(!m_most_recent_block_mutex);
```
Comment (MarcoFalke) [inline src/net_processing.cpp:476]: I wonder how useful it is to annotate interface functions with member mutexes. The only allowed caller can't see the mutex, as it is private.

**Review by maflcko — APPROVED** [review]
lgtm

File: src/net_processing.cpp (hunk 1 above)
Comment (hebasto) [reply src/net_processing.cpp:476]: An annotation is not for a caller, rather for Clang Thread Safety Analysis C++ extension.

**Comment by ajtowns** [comment]
ACK 2b3373c1520aa0b41277cd89956224e08cbd79dd

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25175/commits/2b3373c1520aa0b41277cd89956224e08cbd79dd

File: src/net_processing.cpp (hunk 1 above)
Comment (ajtowns) [reply src/net_processing.cpp:476]: (An external caller (accessing via the virtual/override) can't see the mutex, but an in-class caller could)

Comment (MarcoFalke) [reply src/net_processing.cpp:476]: Yeah, I was just thinking that it would be incredibly unlikely or even wrong to even have an in-class caller.

**Pull request merged**